### ISSUE TITLE
Experimental feature failure: ReadDataDiff unequal configs

### ISSUE THREAD

**Issue by jtopper** [hashicorp/terraform#10075]
Sharing the output from an experimental feature failure, as requested.

### Terraform Version

`Terraform v0.7.10`

### Affected Resource(s)

aws_iam_policy_document

### Terraform Configuration Files

This is too big and sensitive to provide publicly, but please ask if you need to see it and I'll share privately.

### Output

https://gist.github.com/jtopper/4aed431bb5422c71d389fe21267c0a26

**Comment by mitchellh** [comment]
Boom, found the root cause for this thanks to a Terraform Enterprise run a customer hit. Fix in #10118.

**Comment by Jamie-BitFlight** [comment]
I'm actually experiencing this in terraform 0.10.7
```
data.aws_iam_policy_document.ddslevel3role - *terraform.NodeValidatableResourceInstance
2017/10/20 11:23:39 [DEBUG] Resource state not found for "data.aws_iam_policy_document.ddslevel2role": data.aws_iam_policy_document.ddslevel2role
2017/10/20 11:23:39 [TRACE] root: eval: *terraform.EvalInterpolate
2017/10/20 11:23:39 [DEBUG] ReferenceTransformer: "data.aws_iam_policy_document.ddslevel2role" references: []
2017/10/20 11:23:39 [ERROR] root: eval: *terraform.EvalValidateResource, err: Warnings: []. Errors: [statement.0.principals.0: expected object, got string]
2017/10/20 11:23:39 [ERROR] root: eval: *terraform.EvalSequence, err: Warnings: []. Errors: [statement.0.principals.0: expected object, got string]
2017/10/20 11:23:39 [TRACE] [walkValidate] Exiting eval tree: data.aws_iam_policy_document.ddslevel2role
2017/10/20 11:23:39 [ERROR] root: eval: *terraform.EvalValidateResource, err: Warnings: []. Errors: [statement.0.principals.0: expected object, got string statement.0.principals.1: expected object, got string statement.0.principals.2: expected object, got string statement.0.principals.3: expected object, got string statement.0.principals.4: expected object, got string statement.0.principals.5: expected object, got string statement.0.principals.6: expected object, got string statement.0.principals.7: expected object, got string]
2017/10/20 11:23:39 [ERROR] root: eval: *terraform.EvalSequence, err: Warnings: []. Errors: [statement.0.principals.0: expected object, got string statement.0.principals.1: expected object, got string statement.0.principals.2: expected object, got string statement.0.principals.3: expected object, got string statement.0.principals.4: expected object, got string statement.0.principals.5: expected object, got string statement.0.principals.6: expected object, got string statement.0.principals.7: expected object, got string]
2017/10/20 11:23:39 [TRACE] [walkValidate] Exiting eval tree: data.aws_iam_policy_document.ddslevel3role
2017/10/20 11:23:39 [TRACE] dag/walk: walking "aws_iam_role.ddslevel2role"
2017/10/20 11:23:39 [DEBUG] Resource state not found for "aws_iam_role.ddslevel2role": aws_iam_role.ddslevel2role
2017/10/20 11:23:39 [DEBUG] Resource state not found for "aws_iam_role.ddslevel3role": aws_iam_role.ddslevel3role
2017/10/20 11:23:39 [TRACE] Graph after step *terraform.AttachStateTransformer:

aws_iam_role.ddslevel2role - *terraform.NodeValidatableResourceInstance
2017/10/20 11:23:39 [TRACE] Graph after step *terraform.AttachStateTransformer:

aws_iam_role.ddslevel3role - *terraform.NodeValidatableResourceInstance
2017/10/20 11:23:39 [DEBUG] Resource state not found for "aws_iam_role_policy_attachment.ddslevel2role[0]": aws_iam_role_policy_attachment.ddslevel2role[0]
2017/10/20 11:23:39 [DEBUG] Resource state not found for "aws_iam_role_policy_attachment.ddslevel2role[1]": aws_iam_role_policy_attachment.ddslevel2role[1]
2017/10/20 11:23:39 [DEBUG] ReferenceTransformer: "aws_iam_role_policy_attachment.ddslevel2role[0]" references: []
2017/10/20 11:23:39 [DEBUG] ReferenceTransformer: "aws_iam_role_policy_attachment.ddslevel2role[1]" references: []
2017/10/20 11:23:39 [DEBUG] Resource state not found for "aws_iam_role_policy_attachment.ddslevel3role[2]": aws_iam_role_policy_attachment.ddslevel3role[2]
2017/10/20 11:23:39 [DEBUG] Resource state not found for "aws_iam_role_policy_attachment.ddslevel3role[3]": aws_iam_role_policy_attachment.ddslevel3role[3]
2017/10/20 11:23:39 [DEBUG] Resource state not found for "aws_iam_role_policy_attachment.ddslevel3role[0]": aws_iam_role_policy_attachment.ddslevel3role[0]
2017/10/20 11:23:39 [DEBUG] ReferenceTransformer: "aws_iam_role_policy_attachment.ddslevel3role[1]" references: []
2017/10/20 11:23:39 [DEBUG] ReferenceTransformer: "aws_iam_role_policy_attachment.ddslevel3role[2]" references: []
2017/10/20 11:23:39 [DEBUG] ReferenceTransformer: "aws_iam_role_policy_attachment.ddslevel3role[3]" references: []
2017/10/20 11:23:39 [TRACE] root: eval: *terraform.EvalValidateResource
2017/10/20 11:23:39 [DEBUG] ReferenceTransformer: "aws_iam_role_policy_attachment.ddslevel3role[0]" references: []
2017/10/20 11:23:39 [DEBUG] plugin: waiting for all plugin processes to complete...
9 error(s) occurred:

* data.aws_iam_policy_document.ddslevel2role: statement.0.principals.0: expected object, got string
* data.aws_iam_policy_document.ddslevel3role: statement.0.principals.0: expected object, got string
* data.aws_iam_policy_document.ddslevel3role: statement.0.principals.1: expected object, got string
* data.aws_iam_policy_document.ddslevel3role: statement.0.principals.2: expected object, got string
* data.aws_iam_policy_document.ddslevel3role: statement.0.principals.3: expected object, got string
* data.aws_iam_policy_document.ddslevel3role: statement.0.principals.4: expected object, got string
* data.aws_iam_policy_document.ddslevel3role: statement.0.principals.5: expected object, got string
* data.aws_iam_policy_document.ddslevel3role: statement.0.principals.6: expected object, got string
* data.aws_iam_policy_document.ddslevel3role: statement.0.principals.7: expected object, got string
```

```
locals {
  a = "arn:aws:iam::${module.accounts.account_ids["dds"]}:user"
  principal_users = [
  "${local.a}/xxxx",
  "${local.a}/yyyy",
  ]
 dds3_policies = [
 "arn:aws:iam::aws:policy/AmazonEC2FullAccess",
 "arn:aws:iam::aws:policy/AWSSupportAccess",
 "arn:aws:iam::aws:policy/AmazonS3FullAccess",
 "arn:aws:iam::aws:policy/ReadOnlyAccess"
   ]
}
######
resource "aws_iam_role" "ddslevel3role" {
  name = "DDS_Level3"
  assume_role_policy = "${data.aws_iam_policy_document.ddslevel3role.json}"
}

resource "aws_iam_role_policy_attachment" "ddslevel3role" {
    count = "${length(local.dds3_policies)}"
    role       = "${aws_iam_role.ddslevel3role.name}"
    policy_arn = "${local.dds3_policies[count.index]}"
}

data "aws_iam_policy_document" "ddslevel3role" {
  policy_id = "DDSLevel3RolePolicy"
  statement {
    actions = ["sts:AssumeRole"]
    effect = "Allow"
    principals = ["${local.principal_users}"]
    condition {
      test     = "Bool"
      variable = "aws:MultiFactorAuthPresent"
      values = ["true"]
    }
    sid = "ddslevel3role"
  }
}
```
